ddtrace/profiling/profiler.py

```diff
@@ -81,30 +60,115 @@ def __repr__(self):
 ProfilerStatus.RUNNING = ProfilerStatus("running")
 
 
-@attr.s
 class Profiler(object):
     """Run profiling while code is executed.
 
     Note that the whole Python process is profiled, not only the code executed. Data from all running threads are
     caught.
 
-    If no collectors are provided, default ones are created.
-    If no exporters are provided, default ones are created.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self._profiler = _ProfilerInstance(*args, **kwargs)
+
+    def start(self, stop_on_exit=True, profile_children=True):
+        """Start the profiler.
+
+        :param stop_on_exit: Whether to stop the profiler and flush the profile on exit.
+        :param profile_children: Whether to start a profiler in child processes.
+                                 The new profiler object will be stored in the `child` attribute.
+        """
+
+        self._profiler.start()
+
+        if stop_on_exit:
+            atexit.register(self.stop)
+
+        if profile_children:
+            if hasattr(os, "register_at_fork"):
+                os.register_at_fork(after_in_child=self._restart_on_fork)
+            else:
+                LOG.warning(
+                    "Your Python version does not have `os.register_at_fork`. "
+                    "You have to start a new Profiler after fork() manually."
+                )
+
+    def stop(self, flush=True):
+        """Stop the profiler.
+
+        :param flush: Wait for the flush of the remaining events before stopping.
+        """
+        self._profiler.stop(flush)
+
+    def _restart_on_fork(self):
+        # Be sure to stop the parent first, since it might have to e.g. unpatch functions
+        self.stop()
+        self._profiler = self._profiler.copy()
+        self._profiler.start()
+
+    @property
+    def status(self):
+        return self._profiler.status
+
+    @property
+    def service(self):
+        return self._profiler.service
+
+    @property
+    def env(self):
+        return self._profiler.env
+
+    @property
+    def version(self):
+        return self._profiler.version
+
+    @property
+    def tracer(self):
+        return self._profiler.tracer
+
+
+@attr.s
+class _ProfilerInstance(object):
```

Comment:
Hmm is a separate class really needed? It seems like we have to pass through a bunch of the attributes to this class and all that is really needed is to call what's in `__attrs_post_init__()` (all `copy()` does is create a new instance). Could that just be pulled out into a function and called again after fork?

Reply:
I'm not sure what you mean. The user is going to keep the `Profiler` object reference forever, even after a fork. We can't change that reference.

So we need to "update" the `_ProfilerInstance` that does the "real" job behind the scene when the process forks. That's why we keep the original object here, renamed, but allow it to be copied to be started fresh when a process forks.

Reply:
Yeah it'll be kept/copied from the fork, but I don't see the need in changing the reference. All that we end up doing here is reinitializing the class. The only important part I see that needs to be done on fork is to reinitialize the collectors (and maybe the exporters?). We could just do this explicitly instead of having a whole new class I think?

I'm not too concerned but it just seemed a bit off that we're passing around `*args, **kwargs` and proxying a bunch of properties.

Reply:
Tweaking the object and doing "re-init" is usually a good way of having bugs (see the `Tracer`…) because you have two different code paths to do your initialization (one for init, one for re-init).

This approach of doing a `copy()` has the upside of being clearer (yay functional programming. You just throw an object because you don't need it anymore (it's dead after a fork) and creates a new one by copying the existing one, which re-use the same init code path.

It also has the upside of splitting the profiler code and the fork handling in two different abstractions, making it easier to test only a part of the code at a time.

So yes I really stand by this design, unless missing something. :)